Add Grindstone experience reward API

Sytm opened the pull request:
The only "difference" to Vanilla behavior is when then experience reward will be determined, but otherwise it is left as is.

This supersedes #9762

Commit:
Add Grindstone experience reward API

lynxplay reviewed (commented):
After a bit of thought, I think moving the time of computation and "ownership" of the experience value from the slot to the inventory is a tad of a dangerous game.

The experience does not exist, we move the time of computation here to a place that may potentially now be out of sync with the actual computation down the line.

Having the inventory "manage" the exp also allows plugins to magically change this value outside of any events being called if they simply keep a reference to it,  leading to potentially more disjoint between expected exp and rewarded one.

This is by far not a hard no from me, I just think the API suggestions needs a lot more evaluation.

Commit:
Make experience award only modifiable from the PrepareGrindstoneEvent

(I am aware of the fact that this patch modifies code from another patch, I would merge them if this is deemed a better solution)
